File: frontend/src/lib/paper.ts
/**
 * Paper size specs — the single source of truth for pagination + export sizing.
 * Mirrors `pkg/export/paper.go`. Values use the CSS reference pixel (96dpi) so
 * `1in = 96px` and `1mm = 96/25.4 px`, matching how the headless renderer lays
 * out `mm`-based CSS and how Chromium reports PDF paper size in inches.
 */

export type PaperSizeName = 'A4' | 'Letter'
export type Orientation = 'portrait' | 'landscape'

export interface PaperSpec {
  name: PaperSizeName
  /** Portrait dimensions in millimetres (landscape swaps them). */
  mmW: number
  mmH: number
  /** Dimensions in CSS pixels (96dpi). */
  pxW: number
  pxH: number
  /** Dimensions in inches (for headless print APIs). */
  inchW: number
  inchH: number
}

/** CSS reference DPI: 1in = 96px. */
export const MM_TO_PX = 96 / 25.4

const BASE: Record<PaperSizeName, { mmW: number; mmH: number }> = {
  A4: { mmW: 210, mmH: 297 },
  Letter: { mmW: 215.9, mmH: 279.4 },
}

export const DEFAULT_PAPER_SIZE: PaperSizeName = 'A4'
export const DEFAULT_ORIENTATION: Orientation = 'portrait'

/** Normalizes an arbitrary paper-size string to a known spec (fallback A4). */
export function normalizePaperSize(value: string | null | undefined): PaperSizeName {
  if (value === 'Letter' || value === 'letter') return 'Letter'
  return 'A4'
}

function isLandscape(orientation: string | null | undefined): boolean {
  return orientation === 'landscape'
}

/** Resolves a full spec from a paper-size name and orientation. */
export function resolvePaper(
  size?: string | null,
  orientation?: string | null,
): PaperSpec {
  const name = normalizePaperSize(size)
  const base = BASE[name]
  const landscape = isLandscape(orientation)
  const mmW = landscape ? base.mmH : base.mmW
  const mmH = landscape ? base.mmW : base.mmH
  return {
    name,
    mmW,
    mmH,
    pxW: Math.round(mmW * MM_TO_PX),
    pxH: Math.round(mmH * MM_TO_PX),
    inchW: round2(mmW / 25.4),
    inchH: round2(mmH / 25.4),
  }
}

function round2(v: number): number {
  return Math.round(v * 100) / 100
}

/** Default A4-portrait spec. */
export const DEFAULT_PAPER: PaperSpec = resolvePaper()

/**
 * A4-portrait shortcuts kept for existing call sites (e.g. the preview panel's
 * chrome width). Prefer {@link resolvePaper} for anything template-dependent.
 */
export const A4_W = DEFAULT_PAPER.pxW
export const A4_H = DEFAULT_PAPER.pxH

File: pkg/template_export/paper.go
package template_export

import (
	"math"
	"regexp"
	"strings"
)

// PaperSpec 描述用于分页与导出的物理纸张规格。
//
// 各字段取值与前端 frontend/src/lib/paper.ts 保持一致。px 采用 CSS 参考像素
// （96dpi），即 1in = 96px、1mm = 96/25.4 px——与无头 Chromium 排版 mm 单位
// CSS 的方式，以及它以英寸上报 PDF 纸张尺寸的方式一致。
type PaperSpec struct {
	Name  string
	MmW   float64
	MmH   float64
	PxW   int
	PxH   int
	InchW float64
	InchH float64
}

var (
	// PaperA4 是 A4 纵向规格（210 × 297 mm）。
	PaperA4 = makePaper("A4", 210, 297)
	// PaperLetter 是 Letter 纵向规格（8.5 × 11 in）。
	PaperLetter = makePaper("Letter", 215.9, 279.4)
)

// mmToPx 是毫米到 CSS 参考像素的换算系数（96dpi）。
const mmToPx = 96 / 25.4

// makePaper 由毫米尺寸推导出像素与英寸尺寸，构造完整的纸张规格。
func makePaper(name string, mmW, mmH float64) PaperSpec {
	return PaperSpec{
		Name:  name,
		MmW:   mmW,
		MmH:   mmH,
		PxW:   int(math.Round(mmW * mmToPx)),
		PxH:   int(math.Round(mmH * mmToPx)),
		InchW: round2(mmW / 25.4),
		InchH: round2(mmH / 25.4),
	}
}

// round2 保留两位小数，用于英寸尺寸取整。
func round2(v float64) float64 {
	return math.Round(v*100) / 100
}

// ResolvePaper 按纸张名称与横向标记返回对应规格。
// 名称未知时回退到 A4；landscape 为 true 时交换宽高。
func ResolvePaper(name string, landscape bool) PaperSpec {
	var base PaperSpec
	switch strings.ToLower(name) {
	case "letter":
		base = PaperLetter
	default:
		base = PaperA4
	}
	if !landscape {
		return base
	}
	return PaperSpec{
		Name:  base.Name,
		MmW:   base.MmH,
		MmH:   base.MmW,
		PxW:   base.PxH,
		PxH:   base.PxW,
		InchW: base.InchH,
		InchH: base.InchW,
	}
}

// 用于从 HTML 中提取前端标注的纸张规格属性。
var (
	paperSizeRe   = regexp.MustCompile(`data-paper-size="([^"]*)"`)
	orientationRe = regexp.MustCompile(`data-orientation="([^"]*)"`)
)

// PaperFromHTML 读取前端渲染器标注在 .resume-page 元素上的纸张规格。
// 未找到相应属性时回退到 A4 纵向。
func PaperFromHTML(html string) PaperSpec {
	name := ""
	orientation := ""
	if m := paperSizeRe.FindStringSubmatch(html); m != nil {
		name = m[1]
	}
	if m := orientationRe.FindStringSubmatch(html); m != nil {
		orientation = m[1]
	}
	return ResolvePaper(name, orientation == "landscape")
}
